File: gamit/message/message.py
__author__ = 'mahanzhou'

from gamit.serialize.serializer import Serializer
from  gamit.serialize.datatype import RmiDataType

class MessageRegisteredError(Exception):
    pass

class MessageNotRegisteredError(Exception):
    pass

class MessageBlock:
    _msgMap = {}

    @classmethod
    def register(cls, messageType):
        #if messageType.__name__ in cls.__msgMap:
        #    raise MessageRegisteredError()
        #
        cls._msgMap[messageType.__name__] = messageType

    def __init__(self, dataType, toIdList=None, data=None):
        if isinstance(dataType, int):
            self.command = dataType
            self.toIdList = toIdList
            self.data = data
            self._os = Serializer()
            self._write()
        else:
            self._is = dataType
            self._read()

    def _read(self):
        self.command = self._is.readInt() # 1 command
        self.toIdList = []
        idSize = self._is.readInt()       # 2 toIdList
        for dummy in range(0, idSize):
            id = self._is.readInt()
            self.toIdList.append(id)
        name = self._is.readString()      # 3 message name
        if name not in self._msgMap:
            raise MessageNotRegisteredError()
        self.data = self._msgMap[name]()  # 4 data
        self.data._read(self._is)

    def _write(self):
        self._os.startToWrite()
        self._os.writeByte(RmiDataType.MessageBlock)
        self._os.writeInt(self.command)  # 1 command
        idSize = len(self.toIdList)       # 2 toIdList
        self._os.writeInt(idSize)
        for id in self.toIdList:
            self._os.writeInt(id)
        self._os.writeString(self.data.__class__.__name__) # 3 message name
        self.data._write(self._os)      # 4 data

    def getOsBuffer(self):
        return self._os.getBuffer()

File: cpp/Source/gamit/message/Message.cpp
#include "Message.h"

using namespace gamit;

MessageBlock::MapCreateFunc MessageBlock::_mapCreateFunc;

MessageBlock::MessageBlock(CSerializer & __is)
:_command()
,_messageBase()
, __os()
{
	__is.read(_command);

	int size = 0;
	__is.read(size);

	// toIdList
	for (int i = 0; i < size; i++)
	{
		int dummy = 0;
		__is.read(dummy);
	}

	std::string msgName;
	__is.read(msgName);

	_messageBase = createMessageBase(msgName);
	if (nullptr != _messageBase)
	{
		_messageBase->__read(__is);
	}
}

MessageBlock::MessageBlock(int command, const MessageBasePtr & msgBase)
:_command(command)
,_messageBase(msgBase)
,__os()
{
	__os.startToWrite();
	__os.write(byte_t(ERmiType::MessageBlock));
	__os.write(_command);
	__os.write(int(0)); //no toIdList
	__os.write(_messageBase->__name());

	if (nullptr != _messageBase)
	{
		_messageBase->__write(__os);
	}
}

const std::string & MessageBlock::getBuffer()
{
	return __os.getBuffer();
}

MessageBasePtr MessageBlock::createMessageBase(const std::string & msgName)
{
	auto found = _mapCreateFunc.find(msgName);

	if (found == _mapCreateFunc.end())
	{
		return nullptr;
	}

	return found->second();
}

void MessageBlock::registCreateFunc(const std::string & msgName, CreateFunc func)
{
	_mapCreateFunc[msgName] = func;
}
